x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -130,6 +134,23 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeMap(metadata);
     }
 
+    public boolean sameUserAs(Authentication other) {
+        if (AuthenticationType.API_KEY == getAuthenticationType() && AuthenticationType.API_KEY == other.getAuthenticationType()) {
+            assert getUser().principal().equals(other.getUser().principal()) :
+                "The same API key ID cannot be attributed to two different usernames";
+            return getMetadata().get(API_KEY_ID_KEY).equals(other.getMetadata().get(API_KEY_ID_KEY));
+        }
+        if (false == getUser().principal().equals(other.getUser().principal())) {
+            return false;
+        }
+        final RealmRef thisRealm = getSourceRealm();
+        final RealmRef otherRealm = other.getSourceRealm();
+        if (FileRealmSettings.TYPE.equals(thisRealm.getType()) || NativeRealmSettings.TYPE.equals(thisRealm.getType())) {
+            return thisRealm.getType().equals(otherRealm.getType());
+        }
+        return thisRealm.getName().equals(otherRealm.getName()) && thisRealm.getType().equals(otherRealm.getType());
+    }
+
```

Comment:
I think any user should have access to the resources owned by the anonymous user.
If you agree, I propose we handle it in a follow-up.

Reply:
I am not sure about this. I don't think it should. We never allow one user accessing another user's resource without explicitly granting permissions. For a contrived example, say the anonymous user can `manage_own_api_key`. Currently we don't allow any other user to manage anonymous user's API keys. I think it is better and easier to understand if we don't make this happen automatically. I am aware that any roles granted to the anonymous user are automatically granted to any other user. But I think it is a bit different because it is not about assuming other user's identity.

Reply:
Okay, makes sense. Thanks for considering it!